src/main/java/com/google/api/codegen/util/ProtoParser.java

```diff
@@ -175,32 +184,39 @@ private HttpRule getHttpRule(Method method) {
       List<ProtoFile> protoFiles, DiagCollector diagCollector) {
     ImmutableMap.Builder<String, ResourceDescriptorConfig> mapBuilder = ImmutableMap.builder();
 
+    // Maps base names to ResourceDescriptors. Used to check redeclarations.
+    Map<String, ResourceDescriptor> localDefs = new LinkedHashMap<>();
+
     // Skip unnecessary file parsing.
     if (!enableProtoAnnotations) return mapBuilder.build();
 
     for (ProtoFile protoFile : protoFiles) {
+      List<ResourceDescriptor> resourceDescriptors = new ArrayList<>();
 
-      // Maps base names to ResourceDescriptors.
-      Map<String, ResourceDescriptor> localDefs = new LinkedHashMap<>();
+      // Get Resource[Set] definitions from file-level annotations.
+      resourceDescriptors.addAll(getFileLevelResourceDescriptors(protoFile));
 
       // Get Resource[Set] definitions from fields in message types.
-      for (MessageType message : protoFile.getMessages()) {
-        ResourceDescriptor definition = getResourceDescriptor(message);
-        if (definition != null) {
-          if (localDefs.put(definition.getType(), definition) != null) {
-            diagCollector.addDiag(
-                Diag.error(
-                    SimpleLocation.TOPLEVEL,
-                    "Multiple ResourceDescriptor defintions with the type"
-                        + " %s are defined in proto file %s. Values for type must be unique.",
-                    definition.getType(),
-                    protoFile.getFullName()));
-            continue;
-          }
-          ResourceDescriptorConfig config =
-              ResourceDescriptorConfig.from(definition, message.getFile());
-          mapBuilder.put(config.getUnifiedResourceType(), config);
+      protoFile
```

Comment:
(stylistic and subjective, a suggestion, not a requirement): this code looks more complicated than it should.

I think the reason why you had to introduce an intermediate collection (`resourceDescriptors`) is because we needed to iterate in same way over 2 sets of data, retrieved differently: 
set1: fileLevelResourceDescriptors (new in this PR)
set2: message-leel resource descriptors (old stuff)

To do that we here introduce an intermediate collection (`resourceDescriptors`) and then mix two styles (for loop and java8 streams) to process that. 

WDYT about just moving the processing logic in a private method and simply iterate over 2 sets of data twice, calling that private method (in either way: java8 streams or for loop)? It will achieve same result, will keep code style more consistent, kind of more efficient (no need to have intermediate collection) and easier to understand (it will be immediately clear that we just do same thing over file descriptors and then over message descriptors).

Reply:
That makes sense! I'll update it.